src/target/riscv/riscv-013.c

```diff
@@ -2471,6 +2471,33 @@ static int read_memory_progbuf(struct target *target, target_addr_t address,
 
 	memset(buffer, 0, count*size);
 
+	uint64_t mstatus = 0;
+	uint64_t mstatus_old = 0;
+	if (progbufsize >= 4) {
+		/* Read DCSR */
+		uint64_t dcsr;
+		if (register_read(target, &dcsr, GDB_REGNO_DCSR) != ERROR_OK)
+			return ERROR_FAIL;
+
+		/* Read and save MSTATUS */
+		if (register_read(target, &mstatus, GDB_REGNO_MSTATUS) != ERROR_OK)
+			return ERROR_FAIL;
+		mstatus_old = mstatus;
+
+		/* If we come from m-mode with mprv set, we want to keep mpp */
+		if (get_field(dcsr, DCSR_PRV) < 3) {
+			/* MPP = PRIV */
+			mstatus = set_field(mstatus, MSTATUS_MPP, get_field(dcsr, DCSR_PRV));
+
+			/* MPRV = 1 */
+			mstatus = set_field(mstatus, MSTATUS_MPRV, 1);
+
+			/* Write MSTATUS */
+			if (register_write_direct(target, GDB_REGNO_MSTATUS, mstatus) != ERROR_OK)
```

Comment:
We should only do this write if it will actually change the value of mstatus.

Reply:
Done